home/tests.py

```diff
@@ -193,3 +197,105 @@ def test_response_context(self, response):
+
+        @pytest.mark.parametrize("invalid_Input, exception", [
+            ((10, None, "Question in Math", 'How much is it 1+1?',
+             datetime(2022, 4, 7, 12, 53, 29, 4, tzinfo=pytz.UTC), 1, 4, Grade.GRADE7, 1, 10, False), ValidationError),
+            # User didn't enter title
+            ((10, 1, None,
+             datetime(2022, 4, 7, 12, 53, 29, 4, tzinfo=pytz.UTC), 1, 4, Grade.GRADE7, 1, 10, False), TypeError),
+            # User did'nt enter content
+            ((10, 1, "Question in Math", 'How much is it 1+1?', "", 1, 4, Grade.GRADE7, 1, 10, False), ValidationError),
+            # The question does not contain date
+            ((10, 1, "Question in Math", 'How much is it 1+1?',
+             datetime(2022, 4, 7, 12, 53, 29, 4, tzinfo=pytz.UTC), "", 4, Grade.GRADE7, 1, 10, False), ValidationError),
+            # User didn't choose subject
+            ((10, 1, "Question in Math", 'How much is it 1+1?',
+             datetime(2022, 4, 7, 12, 53, 29, 4, tzinfo=pytz.UTC), 1, 4, "", 1, 10, False), ValidationError),
+            # User did'nt choose grade
+            ((10, 1, "Question in Math", 'How much is it 1+1?',
+             datetime(2022, 4, 7, 12, 53, 29, 4, tzinfo=pytz.UTC), 1, 4, Grade.GRADE7, 1, -10, False), ValidationError),
+            # User chose negative book number
+            ((10, 1, "Question in Math", 'How much is it 1+1?',
+             datetime(2022, 4, 7, 12, 53, 29, 4, tzinfo=pytz.UTC),
+             1, 4, Grade.GRADE7, 1, 99990, False), ValidationError),
+            # User entered to high book number
+        ])
+        # checks if the question was not inserted
+        @pytest.mark.django_db
+        def test_add_invalid_question(self, invalid_Input, exception):
+
+            with pytest.raises(exception):
+                assert Question(*invalid_Input).clean_fields()
+
+        # checks default value in is_edited
+        @pytest.mark.django_db
+        def test_default_is_edited(self, question):
+
+            assert question.is_edited is False
+
+        # checks default value in book_page once the user didn't insert a book page
+        @pytest.mark.django_db
+        def test_default_book_page(self, question):
+
+            assert question.book_page is None
+
+        # checks default value in book once the user didn't insert a book
+        @pytest.mark.django_db
+        def test_default_book(self, question):
+
+            assert question.book is None
+
+        # checks default value in sub_subject once the user didn't insert a sub-subject
+        @pytest.mark.django_db
+        def test_default_sub_subject(self, question):
+
+            assert question.sub_subject is None
+
+    class TestViews:
+        @pytest.mark.django_db
+        def test_uses_question_form(self, client):
+            response = client.get('/explore/new_question')
+
+            assert response.status_code == 200
+            assert isinstance(response.context['form'], type(QuestionForm))
```

Comment:
missing a test for the various flows of posting data - when proper data is posted, and when bad data is posted...

Reply:
I made these tests at TestDatabaseInsertions class, why do I have to do it also in TestViews class? what is the difference?